### PR TITLE
Improve breeding and validation of fodder references

### LINKED ISSUE #212: Variables are not set in fodder gene
Trying to set a samples using variables, but password cannot be set in referenced starter food.
Locks to be a bug or I missed something in configuration:

catlet: 

``` yaml
parent: dbosoft/ubuntu-22.04/starter

memory:
  startup: 1024

variables:
  - name: password
    required: true
    secret: true

fodder: 
 - source: gene:dbosoft/starter-food:linux-starter
   variables: 
   - name: password
     value: {{ password }}
```

Variable password is not set in starter food and keeps default value "admin".
Also tried source gene:dbosoft/starter-food/1.1:linux-starter without success.

For reference, the updated linux-starter 1.1 contains following variables: 

 
``` yaml
name: linux-starter

variables:
  - name: username
    value: admin
    required: true
  - name: password
    required: true
    secret: true
    value: admin
  - name: lockPassword
    type: boolean
    value: false    
  - name: sshPublicKey
    required: false

fodder:
- name: admin-linux
  type: cloud-config
  secret: true
  content: |
    users:
    - name: {{ username }}
      plain_text_passwd: {{ password }}
      group: admin
      lock_passwd: {{ lockPassword }}
      sudo: 'ALL=(ALL) NOPASSWD: ALL'
      ssh_authorized_keys:
      - {{ sshPublicKey }}

```

[AB#679](https://dev.azure.com/dbosoft/eae53133-585f-4d55-8e1f-5cba6fd8ce3b/_workitems/edit/679)

### PR DESCRIPTION
**Pull request by ChristopherMann**
This PR improves the breeding and validation of fodder references. Important changes:
- The reserved gene name `catlet` is now used in the source for fodder which was taken from the parent. This source is for informational purposes only. We no longer attempt to resolve these sources like other fodder genes.
- During breeding and feeding, we now resolve all gene identifiers before filtering the fodder according to the removed flag.

Closes #212

### CODE REVIEW

File: src/core/src/Eryph.Core/Genetics/CatletBreeding.cs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    /// </summary>
+    public static Either<Error, CatletConfig> Breed(
+        CatletConfig parentConfig,
+        CatletConfig childConfig) =>
+        from cpu in BreedCpu(parentConfig.Cpu, childConfig.Cpu)
+        from memory in BreedMemory(parentConfig.Memory, childConfig.Memory)
+        from drives in BreedDrives(Seq(parentConfig.Drives), Seq(childConfig.Drives))
+        from networkAdapters in BreedNetworkAdapters(Seq(parentConfig.NetworkAdapters), Seq(childConfig.NetworkAdapters))
+        from networks in BreedNetworks(Seq(parentConfig.Networks), Seq(childConfig.Networks))
+        from variables in BreedVariables(Seq(parentConfig.Variables), Seq(childConfig.Variables))
```
Comment (fw2568) [inline src/core/src/Eryph.Core/Genetics/CatletBreeding.cs:26]: Not sure if missing variables actually stop the breeding process - I called the API first with an older compute client that knows nothing about variables - and it did not fail. 

Catlet used:

``` yaml
parent: dbosoft/ubuntu-22.04/starter

memory:
  startup: 1024

variables:
  - name: password
    required: true
    secret: true

fodder: 
 - source: gene:dbosoft/starter-food:linux-starter
   variables: 
   - name: password
     value: "{{ password }}"
```

**Review by fw2568 — CHANGES_REQUESTED** [review]
I'm still not able to breed catlet with the config of #212, but maybe I missed something.

My local config is modified to use dbosoft/starter-food/1.1 for dbosoft/starter-food/latest, which contains variables for password and user.

File: src/core/src/Eryph.Core/Genetics/CatletBreeding.cs (hunk 1 above)
Comment (ChristopherMann) [reply src/core/src/Eryph.Core/Genetics/CatletBreeding.cs:26]: Missing variables will stop the creation of the catlet, but this only happens in the `UpdateCatletSaga` when the fodder genes are fed to the catlet and the variables are substituted.

When you used the old compute client, it did not include the variables both on catlet level and on fodder level. The config, which the server received, looked like this:
```
parent: dbosoft/ubuntu-22.04/starter

memory:
  startup: 1024

fodder: 
 - source: gene:dbosoft/starter-food:linux-starter
```
This is a valid config as the `linux-starter` food contains (default) values for all variables.

Comment (fw2568) [reply src/core/src/Eryph.Core/Genetics/CatletBreeding.cs:26]: true, forget about the client side parsing.

File: src/modules/src/Eryph.Modules.Controller/Compute/UpdateCatletSaga.cs [hunk 2]
```diff
@@ -2,287 +2,373 @@
...
+    CatletMetadata metadata,
+    HashMap<GeneSetIdentifier, GeneSetIdentifier> resolvedGeneSets,
+    HashMap<GeneSetIdentifier, CatletConfig> parentConfigs) =>
+    from resolvedConfig in CatletGeneResolving.ResolveGeneSetIdentifiers(
+            config, resolvedGeneSets)
+        .MapLeft(e => Error.New("Could not resolve genes in catlet config.", e))
+    from breedingResult in CatletPedigree.Breed(
+            config, resolvedGeneSets, parentConfigs)
+        .MapLeft(e => Error.New("Could not breed the catlet.", e))
+    // After the catlet was created, the fodder and variables can no longer be changed.
```
Comment (fw2568) [inline src/modules/src/Eryph.Modules.Controller/Compute/UpdateCatletSaga.cs:303]: The user experience for update is a bit inconsistant in this case. Update-Catlet asks for variables, even if not supported to change them. So we should remove variable input for update catlet (the only use case I see here is that catlet was not started in between, but this is something we could ignore as edge.

**Review by fw2568 — CHANGES_REQUESTED** [review]
see comment for update catlet

File: src/modules/src/Eryph.Modules.Controller/Compute/UpdateCatletSaga.cs [hunk 3]
```diff
@@ -2,287 +2,373 @@
...
+                AgentName = Data.Data.AgentName,
+                NewStorageId = idGenerator.CreateId(),
+                MachineMetadata = metadata.ValueUnsafe(),
+                MachineNetworkSettings = r.NetworkSettings
             });
-        }
+        });
+    }
+
+    public Task Handle(OperationTaskStatusEvent<UpdateCatletVMCommand> message)
```
[the remark this answers is not in the document]
Comment (fw2568) [reply src/modules/src/Eryph.Modules.Controller/Compute/UpdateCatletSaga.cs:210]: solved with the force-push

File: src/modules/src/Eryph.Modules.Controller/Compute/UpdateCatletSaga.cs (hunk 2 above)
Comment (ChristopherMann) [reply src/modules/src/Eryph.Modules.Controller/Compute/UpdateCatletSaga.cs:303]: I have removed the variable input from the `Update-Catlet` Cmdlet.

**Review by fw2568 — APPROVED** [review]

**Pull request merged**